Extract AIUtils and HarvesterManager from HackyAI

reaperrr opened the pull request:
As another step towards making HackyAI more modular (and making these modules accessible by other AI types), I've extracted a few utils and harvester management to separate files.

reaperrr committed:
Move some stuff from HackyAI to new AIUtils

reaperrr committed:
Extract a HarvesterManager from HackyAI

This takes action when AI harvesters don't find ore near the base or became idle for some other reason.

reaperrr commented:
Updated.

PeterAntal commented on OpenRA.Mods.Common/AI/HarvesterManager.cs:
```diff
@@ -0,0 +1,98 @@
+			claimLayer = world.WorldActor.TraitOrDefault<ResourceClaimLayer>();
+		}
+
+		CPos FindNextResource(Actor actor, Harvester harv)
```
Devil's advocate - Is there a reason FindNextResource method doesn't belong as a member of Harvester? (param overhead might be)

I ask, as conceptually, it seems like this is operating in terms of next resource from pov of that harvester.

reaperrr replied:
I didn't write this part back when it was introduced, but iirc it's
> containment of AI responsability

abcdefg30 commented on OpenRA.Mods.Common/AI/HarvesterManager.cs:
```diff
@@ -0,0 +1,98 @@
+
+		public void GiveOrdersToIdleHarvesters(List<Actor> activeUnits)
+		{
+			if (resLayer == null && resLayer.IsResourceLayerEmpty)
```
This `&&` wants to be an `||`.

abcdefg30 commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -808,9 +747,7 @@ void FindNewUnits(Actor self)
 			foreach (var a in newUnits)
 			{
-				if (a.Info.HasTraitInfo<HarvesterInfo>())
-					QueueOrder(new Order("Harvest", a, false));
```
This change isn't resembled in the new manager. Is it save to drop this order?

reaperrr replied:
It should be. I can't imagine any situation in which the HarvManager won't catch and give an order to an idle harvester.
And as long as there are resources in the vincinity of the production facility that spawned it, the harvesters' own FindResources will trigger.

So as far as I can tell, this Order was redundant.

GraionDilach replied:
I've seen it happenning that since AI new units reach the rally point with an attack-move sometimes the AI used new war miners in base defence due to the attack move. But these units went to harvest after they defeated the threat in the base and started idling thenafter, so I presume this'll be fine as well.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/AI/AIUtils.cs:
```diff
@@ -0,0 +1,51 @@
+		{
+			var cells = world.ActorsHavingTrait<T>().Where(a => a.Owner == player);
+
+			// TODO: Properly check building foundation rather than 3x3 area.
```
This comment appears to be wrong?

GraionDilach replied:
It isn't wrong.  The tiles are enumerated in 3x3 blocks (since it checks a tile and all it's neighbours), ignoring the actual foundation.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -656,8 +619,7 @@ void AssignRolesToIdleUnits(Actor self)
 				assignRolesTicks = Info.AssignRolesInterval;
-				if (resLayer != null && !resLayer.IsResourceLayerEmpty)
-					GiveOrdersToIdleHarvesters();
+				harvManager.GiveOrdersToIdleHarvesters(activeUnits);
```
Can we rename this to `AssignRoles` or `Tick` or something else a bit more generic?

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -839,9 +748,7 @@ void FindNewUnits(Actor self)
-				if (a.Info.HasTraitInfo<HarvesterInfo>())
-					QueueOrder(new Order("Harvest", a, false));
-				else
+				if (!a.Info.HasTraitInfo<HarvesterInfo>())
```
Can we add harvesters to `Info.UnitsCommonNames.ExcludeFromSquads` in the yaml and then drop this explicit check?

GraionDilach replied:
@reaperrr  consider a :+1: from me when this nit is resolved.

pchote reviewed (commented):
A couple more minor nits

reaperrr committed:
Remove HarvesterInfo look-up from HackyAI

reaperrr committed:
HarvManager update rule

reaperrr commented:
Updated.

For the sake of completeness, I added a preliminary update rule, although it doesn't seem to work and I don't know why. Maybe someone can help debugging it.

pchote commented:
Try this: https://gist.github.com/pchote/9d74dd9d120a8a7bedb743d5087778bc

reaperrr commented:
Updated. Tell me when I can remove the test case.

GraionDilach reviewed (approved):
Ohsorry, :+1: overlooked the last commit. ^^

reaperrr commented:
I consider @abcdefg30's +1 still valid, since all I adressed since his approval were pchote's nits and the update rule, so LGTM.
